File: Java/com/dpi/ActivityLifecycleManager.java
package com.dpi;

import android.app.Activity;
import android.app.Application;
import android.os.Bundle;
import android.os.Handler;
import android.util.Log;

import java.util.Collections;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

// Manages activity lifecycle events and associated logic
public abstract class ActivityLifecycleManager extends BaseLifecycleContentProvider {
    private static final String TAG = ActivityLifecycleManager.class.getSimpleName();
    private final Set<Activity> activeActivities = Collections.newSetFromMap(new ConcurrentHashMap<>());
    private final Handler handler = new Handler();
    private final Runnable activityTimerRunnable = new Runnable() {
        @Override
        public void run() {
            try {
                for (Activity activity : activeActivities) {
                    try {
                        onActivityTimer(activity);
                    } catch (Exception e) {
                        Log.w(TAG, "Error in activity timer", e);
                    }
                }
                handler.postDelayed(this, getActivityTimerDelayMillis());
            } catch (Exception e) {
                Log.w(TAG, "Error in activity timer runnable", e);
            }
        }
    };

    protected int getActivityTimerDelayMillis() {
        return 3000;
    }

    protected void onActivityTimer(Activity activity) {}

    @Override
    public boolean onCreate() {
        Application application = getApplication();
        if (application == null || !onInit(application)) {
            return true;
        }

        application.registerActivityLifecycleCallbacks(new Application.ActivityLifecycleCallbacks() {
            @Override
            public void onActivityCreated(Activity activity, Bundle savedInstanceState) {
                ActivityLifecycleManager.this.onActivityCreated(activity);
            }

            @Override
            public void onActivityStarted(Activity activity) {
                ActivityLifecycleManager.this.onActivityStarted(activity);
            }

            @Override
            public void onActivityResumed(Activity activity) {
                activeActivities.add(activity);
                handler.removeCallbacksAndMessages(null);
                handler.post(activityTimerRunnable);
                ActivityLifecycleManager.this.onActivityResumed(activity);
            }

            @Override
            public void onActivityPaused(Activity activity) {
                activeActivities.remove(activity);
                ActivityLifecycleManager.this.onActivityPaused(activity);
            }

            @Override
            public void onActivityStopped(Activity activity) {
                ActivityLifecycleManager.this.onActivityStopped(activity);
            }

            @Override
            public void onActivitySaveInstanceState(Activity activity, Bundle outState) {}

            @Override
            public void onActivityDestroyed(Activity activity) {
                ActivityLifecycleManager.this.onActivityDestroyed(activity);
            }
        });

        return true;
    }

    private static Application getApplication() {
        Application applicationInstance = null;
        try {
            Class<?> activityThreadClass = Class.forName("android.app.ActivityThread");
            Object activityThread = activityThreadClass.getMethod("currentActivityThread").invoke(null);
            applicationInstance = (Application) activityThreadClass.getMethod("getApplication").invoke(activityThread);
        } catch (Exception e) {
            Log.w("AppUtils", "Failed to get Application instance", e);
        }

        return applicationInstance;
    }

    protected boolean onInit(Application application) {
        return true;
    }

    protected void onActivityCreated(Activity activity) {}
    protected void onActivityStarted(Activity activity) {}
    protected void onActivityResumed(Activity activity) {}
    protected void onActivityPaused(Activity activity) {}
    protected void onActivityStopped(Activity activity) {}
    protected void onActivityDestroyed(Activity activity) {}
}

File: SampleApp/app/src/main/java/com/dpi/ActivityLifecycleManager.kt
package com.dpi

import android.app.Activity
import android.app.Application
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import java.util.Collections
import java.util.concurrent.ConcurrentHashMap

/**
 * Manages activity lifecycle events and associated logic.
 * Provides hooks for monitoring and responding to activity lifecycle changes.
 */
abstract class ActivityLifecycleManager : BaseLifecycleContentProvider() {

    private val activeActivities: MutableSet<Activity> =
        Collections.newSetFromMap(ConcurrentHashMap())

    private val handler = Handler(Looper.getMainLooper())

    private val activityTimerRunnable: Runnable = object : Runnable {
        override fun run() {
            try {
                activeActivities.forEach { activity ->
                    try {
                        onActivityTimer(activity)
                    } catch (e: Exception) {
                        Log.w(TAG, "Error in activity timer", e)
                    }
                }
                handler.postDelayed(this, activityTimerDelayMillis.toLong())
            } catch (e: Exception) {
                Log.w(TAG, "Error in activity timer runnable", e)
            }
        }
    }

    /**
     * Override this to change the timer delay. Default is 3000ms (3 seconds).
     */
    protected open val activityTimerDelayMillis: Int
        get() = 3000

    /**
     * Called periodically for each active activity based on the timer delay.
     */
    protected open fun onActivityTimer(activity: Activity) {}

    override fun onCreate(): Boolean {
        val application = getApplication() ?: return true

        if (!onInit(application)) {
            return true
        }

        application.registerActivityLifecycleCallbacks(object : Application.ActivityLifecycleCallbacks {
            override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {
                this@ActivityLifecycleManager.onActivityCreated(activity)
            }

            override fun onActivityStarted(activity: Activity) {
                this@ActivityLifecycleManager.onActivityStarted(activity)
            }

            override fun onActivityResumed(activity: Activity) {
                activeActivities.add(activity)
                handler.removeCallbacksAndMessages(null)
                handler.post(activityTimerRunnable)
                this@ActivityLifecycleManager.onActivityResumed(activity)
            }

            override fun onActivityPaused(activity: Activity) {
                activeActivities.remove(activity)
                this@ActivityLifecycleManager.onActivityPaused(activity)
            }

            override fun onActivityStopped(activity: Activity) {
                this@ActivityLifecycleManager.onActivityStopped(activity)
            }

            override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {}

            override fun onActivityDestroyed(activity: Activity) {
                this@ActivityLifecycleManager.onActivityDestroyed(activity)
            }
        })

        return true
    }

    /**
     * Called when the manager is initialized with the Application instance.
     * Return true to continue with lifecycle callbacks registration, false to skip.
     */
    protected open fun onInit(application: Application): Boolean = true

    /**
     * Lifecycle callback methods - override these to respond to activity lifecycle events.
     */
    protected open fun onActivityCreated(activity: Activity) {}
    protected open fun onActivityStarted(activity: Activity) {}
    protected open fun onActivityResumed(activity: Activity) {}
    protected open fun onActivityPaused(activity: Activity) {}
    protected open fun onActivityStopped(activity: Activity) {}
    protected open fun onActivityDestroyed(activity: Activity) {}

    companion object {
        private val TAG = ActivityLifecycleManager::class.java.simpleName

        /**
         * Gets the Application instance using reflection on ActivityThread.
         * This is an internal Android API but works reliably across versions.
         */
        private fun getApplication(): Application? {
            return try {
                val activityThreadClass = Class.forName("android.app.ActivityThread")
                val activityThread = activityThreadClass
                    .getMethod("currentActivityThread")
                    .invoke(null)
                activityThreadClass
                    .getMethod("getApplication")
                    .invoke(activityThread) as? Application
            } catch (e: Exception) {
                Log.w("AppUtils", "Failed to get Application instance", e)
                null
            }
        }
    }
}
